### ISSUE TITLE
Create validator.nu maker

### ISSUE THREAD

**Issue by rusty1s** [neomake/neomake#534]
I tried to create a html5 linter, because I found the default one to be outdated. This is what I wrote so far following this [description](https://github.com/validator/validator/wiki/Service:-Input:-POST-body):

```vim
let g:neomake_html_html5check_maker = {
  \ 'exe': 'curl',
  \ 'args': [
  \ '-H "Content-Type: text/html; charset=utf-8"',
  \ '--data-binary @' . expand('%:p'),
  \ 'https://validator.w3.org/nu/?out=gnu',
  \ ],
  \ 'errorformat': '%m'
  \ }

let g:neomake_html_enabled_makers=['html5check']
```

But when I open an html document, I get the following error from neomake:

```
curl: option --data-binary @foo.html: is unknown
```

This runs in my local terminal, so what do I do wrong?

**Comment by blueyed** [comment]
Try `'--data-binary', '@'.expand('%:p')` (providing it as two args, not a single one).

**Comment by blueyed** [comment]
Interesting idea btw, please provide a PR when you get it working.

Also do not forget to set the error types, i.e. use `%E` in the beginning if there are only errors.

**Comment by rusty1s** [comment]
Ah, thanks a lot. That removes the error. I also changed `'-H "Content-Type: text/html; charset=utf-8"',` to '-H', '"Content-Type: text/html; charset=utf-8"' because that gave me similar errors.

It didn't work though. Instead I get:

```
<html><body><h1>400 Bad request</h1>
Your browser sent an invalid request.
</body></html>
curl: (3) <url> malformed
```

**Comment by blueyed** [comment]
Where do you get the error?

Have you tried outputting what gets invoked and call that manually?!

Woah, the https://github.com/validator/validator/wiki/Output%3A-GNU#type-descriptor section is not trivial..

It works for me somehow:

In autoload/neomake/makers/ft/html.vim:
```vim
function! neomake#makers#ft#html#html5check() abort
    " Ref: https://github.com/validator/validator/wiki/Output%3A-GNU#type-descriptor
    " TODO: info, non-document-error
    let errorformat = join([
                \   '%f:%l%[:.]%c-%*[0-9:.]: %trror: %m',
                \   '%f:%l%[:.]%c: %trror: %m',
                \   '%f: %trror: %m',
                \   '%E%f:%l%[:.]%c-%*[0-9:.]: %m',
                \   '%E%f:%l%[:.]%c: %m',
                \   '%E: %m',
                \ ], ',')
    return {
                \ 'exe': 'curl',
                \ 'args': ['-s',
                \   '-H', 'Content-Type: text/html; charset=utf-8',
                \   '--data-binary', '@' . expand('%:p'),
                \   'https://validator.w3.org/nu/?out=gnu',
                \ ],
                \ 'errorformat': errorformat
                \ }
endfunction
```

The main iissue here seems to be that the file name is not provided, and there should be some post-processing to set it accordingly.

**Comment by rusty1s** [comment]
Yeah, it works for me too if I edit the neomake source as you do.

**Comment by rstacruz** [comment]
I'd love for this to work! Please continue working on it!

**Comment by blueyed** [comment]
@rstacruz 
You could pick it up from here and create a PR, crediting @rusty1s.
What about that? :)

**Comment by still-dreaming-1** [comment]
@rusty1s When you say the defualt one is outdated, I suspect you might just be using an outdated tidy. Tidy now has a github page where the modern version of it is being developed. You can download installers for different operating systems directly from the github releases page: https://github.com/htacg/tidy-html5/releases

I would be weary of making the validator used here the default one. It uploads your code to an external service to be parsed. I think when people install and start using neomake they assume that by default it is not uploading their private code anywhere...

I don't have enough information about how well they work to have an opinion on if modern tidy works as well as that validator service. I do know that the code for that validor service is available here: https://github.com/validator/validator and it is possible to install it locally and use it locally without uploading your code anywhere, but it does require the Java 8 runtime.

I think as far as a default goes, tidy is a good lightweight default. I only whish I knew how to make neomake use a tidy config file on a per project basis. That way I can use the global config file when no project config is found and then override those with a project config file for projects that need to ignore certain HTML best practices. Right now I am working on integrating tidy into my PhantomJS browser tests so that tidy examines the HTML that was sent to the browser for each test case. That has nothing to do with neomake other than I would like to reuse the same tidy config file with that project in neomake.

**Comment by still-dreaming-1** [comment]
I don't know if this is relevant to neomake or not, but I wanted to provide an update about something I said. I do have an opinion now about tidy vs the W3C Nu Markup Validor purely in terms of how good they are as validators. This is because of the real world experience I have gained trying to use and understand tidy with on one of my projects. While tidy can be used as a validator, it is much more suited to trying to clean-up markup than validate it. When you want to customize its behavior, it does not really have options for customizing the validation behavior even though it has many options for customizing how the cleaned-up markup will generate. Even a modern tidy version gives me the impression it is out of date, as the original author of this issue stated. It does not even recognize the official value for the HTML 4 transitional document type, it wants you to use a different value for that instead...

My experience with the W3C validator has been quite positive. I am now going to attempt to install it locally to use with one of my projects.

**Comment by still-dreaming-1** [comment]
It seems my reason for saying tidy gives me the impression it is out of date was off base. It turns out that the current specification no longer has an HTML 4 transitional document type, so what I was trying to use was invalid. Apparently the DOCTYPE information found here is not up to date: https://www.w3.org/QA/2002/04/valid-dtd-list.html. This website is more up to date with the latest changes in the specification: https://html.spec.whatwg.org/

**Comment by still-dreaming-1** [comment]
Actually some of what I said shows that tidy IS somewhat out of date. Since HTML 4 transitional is an outdated doctype, a good validator would by default show a warning or an error if you try to use it. Instead tidy just has an info message, whose purpose is entirely impossible to figure out, telling you which doctype you used in a way that may or may not imply there is something wrong with that...

The validator.nu does show an error for outdated doctypes. In the past, the online version did not do this, but they recently fixed this, as it was not matching the behavior of running it locally.

So I agree validator.nu is a better default validator. But instead of uploading your code to their website, it should look for a local installation of it. It is very easy to install locally. https://github.com/validator/validator

References:
https://github.com/validator/validator/issues/408
https://github.com/htacg/tidy-html5/issues/466

**Comment by sunnyskc** [comment]
No update on this for a while I guess, but in case anyone is interested, I installed a local copy of validator as still-dreaing-1 suggested sing 
`brew install vnu`
then updated my init.vim to look like this
`let g:neomake_html_enabled_makers = ['vnu']
 let g:neomake_html_vnu_maker = {
         \ 'exe': 'vnu',
         \ 'args':['--errors-only'],
         \ 'errorformat': '%E"file:%f":%l%[:.]%c-%*[0-9:.]: error: %m',
      \}
`
and that seemed to do the trick

**Comment by blueyed** [comment]
@sunnyskc 
Thanks for the input.

The `exe` seems to be different based on you install it, right?
E.g. with https://github.com/svenkreiss/html5validator it would be `html5validator`.

But it should not be a problem to check multiple exes, in case the arguments would be the same.

Do you (or somebody else reading this) want to create a PR to add it?
